### PR TITLE
Fix behaviour for arrays when passed with no value and as the first option

### PR DESCRIPTION
**Pull request by maxrimue**
This fixes [yargs/#383](https://github.com/bcoe/yargs/issues/383#issuecomment-185270189).

If passed as the first option, without grouping and value, an array option would not get defined as `eatArray()` does simply abort if it doesn't see any value passed. This fix makes `eatArray()` assign the array option the default value if no value is provided.

There might be a more elegant solution, though...

### CODE REVIEW

**Comment by bcoe** [comment]
@maxrimue this looks reasonable, what was the behavior before?

**Comment by maxrimue** [comment]
@bcoe The behaviour before this patch is detailed in the linked yargs issue. Basically, if you pass an array as a short option without any other added to it, pass no value, and pass at least one additional option, `eatArray()` is called and does not check for if any value is passed, which means that this:
```
myApp -a --someOption // Where 'a' is an array option
```
gets us `argv.a === undefined` instead of the desired `argv.a === []`

File: index.js [hunk 1]
```diff
@@ -290,7 +290,12 @@ function parse (args, opts) {
   // e.g., --foo apple banana cat becomes ["apple", "banana", "cat"]
   function eatArray (i, key, args) {
     for (var ii = i + 1; ii < args.length; ii++) {
-      if (/^-/.test(args[ii])) break
+      if (/^-/.test(args[ii])) {
+        if (args[ii - 1].replace('-', '') === key) {
```
Comment (bcoe) [inline index.js]: sorry for the slow feedback, finally getting around to merging some things on yargs. I think it might be less fragile to simply use this check:

```js
if (ii === i + 1) {
}
```
ensuring that it's the first iteration of the loop.

Comment (maxrimue) [inline index.js]: @bcoe Won't this also return true in any other iteration? Since `i` gets assigned the value of `ii` at the end of each iteration, and `ii` gets incremented at the start of every iteration, doesn't that mean that `ii === i + 1` will always return true?

I still have passing tests, but I think that's because `yargs-parser` seems to only run arrays through `eatArray()` if they're short grouped and alone as the first option, like this:
```javascript
const yargs = require('yargs');

const argv = yargs
  .array('a')
  .array('b')
  .argv;
```
```
myModule -a -b
```
^ `eatArray()` is only run against `a`. If you call via `myModule -b -a` instead, it's run only against `b`. Is this intentional?

Comment (bcoe) [inline index.js]: whoops, you're right @maxrimue; I think we should add an extra test for this edge-case that I would have broke.

Could we instead do this:

```js
  // if an option is an array, eat all non-hyphenated arguments
  // following it... YUM!
  // e.g., --foo apple banana cat becomes ["apple", "banana", "cat"]
  function eatArray (i, key, args) {
    var start  = i + 1
    for (var ii = i + 1; ii < args.length; ii++) {
      if (/^-/.test(args[ii])) {
        if (ii === start) setArg(key, defaultForType('array'))
        break
      }
      i = ii
      setArg(key, args[ii])
    }

    return i
  }
```

mainly I think it reads clearer, you want to return `[]` if `0` positional arguments follow the starting key?

Comment (maxrimue) [inline index.js]: @bcoe Exactly, since that is what people expect as it's the default value for this type. I like the `start` idea, will update the PR

**Comment by Alphapage** [comment]
Hello,

When do you plan to publish this fix to npm ?
I'm really waiting for this to solve my problems.

Thank you in advance.

**Comment by maxrimue** [comment]
@Alphapage Sorry it took a bit, here we go! We will release a patch to npm soon
